File: 2220-find-all-possible-recipes-from-given-supplies/find-all-possible-recipes-from-given-supplies.cpp
class Solution {
public:
    void buildGraph(vector<string>& recipes, vector<vector<string>>& ingredients, 
    unordered_map<string, int> &ind, unordered_map<string, vector<string>> &adj)
    {
        for(int i = 0; i < recipes.size(); i++)
        {
            string recipe = recipes[i];
            for(string ingredient : ingredients[i])
            {
                if(!adj.count(ingredient))
                adj[ingredient] = vector<string>();

                adj[ingredient].push_back(recipe);
                ind[recipe]++;
            }
        }
    }

    vector<string> findAllRecipes(vector<string>& recipes, vector<vector<string>>& ingredients, vector<string>& supplies) 
    {
        unordered_map<string, vector<string>> adj;
        unordered_map<string, int> ind;
        buildGraph(recipes, ingredients, ind, adj);

        queue<string> q;
        for(string supply : supplies)
        q.push(supply);

        vector<string> res;
        while(!q.empty())
        {
            string node = q.front();
            q.pop();

            if(!adj.count(node))
            continue;

            for(string adjNode : adj[node])
            {
                ind[adjNode]--;
                if(ind[adjNode] == 0)
                {
                    q.push(adjNode);
                    res.push_back(adjNode);
                }
            }
        }
        return res;
    }
};

File: 2220-find-all-possible-recipes-from-given-supplies/find-all-possible-recipes-from-given-supplies.java
class Solution {
    private Map<String, Integer> inDegree;

    private Map<String, Set<String>> buildGraph(String[] recipes, List<List<String>> ingredients) {
        Map<String, Set<String>> adj = new HashMap<>();
        for (int i = 0; i < recipes.length; i++) {
            String recipe = recipes[i];
            inDegree.put(recipe, ingredients.get(i).size());

            for (String ingredient : ingredients.get(i)) {
                adj.computeIfAbsent(ingredient, k -> new HashSet<>()).add(recipe);
            }
        }
        return adj;
    }

    public List<String> findAllRecipes(String[] recipes, List<List<String>> ingredients, String[] supplies) {
        inDegree = new HashMap<>();
        List<String> res = new ArrayList<>();

        // Supplies have in-degree 0
        for (String supply : supplies) {
            inDegree.put(supply, 0);
        }

        // Build the graph
        Map<String, Set<String>> adj = buildGraph(recipes, ingredients);

        // Queue for topological sort
        Queue<String> q = new LinkedList<>();
        for (String item : inDegree.keySet()) {
            if (inDegree.get(item) == 0) {
                q.offer(item);
            }
        }

        while (!q.isEmpty()) {
            String item = q.poll();
            if (!adj.containsKey(item)) continue;

            for (String next : adj.get(item)) {
                inDegree.put(next, inDegree.get(next) - 1);
                if (inDegree.get(next) == 0) {
                    res.add(next);
                    q.offer(next);
                }
            }
        }

        return res;
    }
}
